### ISSUE TITLE
Generic type guard narrows to Unknown

### ISSUE THREAD

**Issue by kuyugama** [astral-sh/ty#2814]
### Summary

I have two types -- `Factory[T]` and `Instance[T]`. I have experimented creating **type-guard functions** and I've discovered that TypeGuard narrowing doesn't work for generics. For example type-guard function that has signature `(Instance[T] | Factory[T]) -> TypeGuard[Factory[T]]` narrows argument to `Factory[Unknown]`

Here is the example:
```python
import typing
from typing_extensions import reveal_type

T = typing.TypeVar("T")


class Instance(typing.Generic[T]):
    pass


class Factory(typing.Generic[T]):
    pass


def isfactory(object: Instance[T] | Factory[T]) -> typing.TypeGuard[Factory[T]]:
    return isinstance(object, Factory)


result = typing.cast(Instance[int] | Factory[int], Instance[int]())

if isfactory(result):
    reveal_type(result)
```

Diagnostics:
```
info[revealed-type]: Revealed type
  --> play2.py:22:17
   |
21 | if isfactory(result):
22 |     reveal_type(result)
   |                 ^^^^^^ `Factory[Unknown]`
   |

Found 1 diagnostic
```

### Version

0.0.17

**Comment by carljm** [comment]
Thanks for the report! This looks like a more general problem in the generic constraint solver, not a problem specific to `TypeGuard`. Here's a more minimal repro:

```py
class A[T]:
    pass

class B[T]:
    pass

def f[T](x: A[T] | B[T]) -> A[T]: ...

def _(x: A[int] | B[int]):
    reveal_type(f(x)) # reveals A[Unknown], should be A[int]
```

This is tracked in #1703

**Comment by carljm** [comment]
(Looking at the original reports both here and in #1703, it does seem like this case is most likely to occur in the real world in use of `TypeIs` or `TypeGuard`, because those are use cases where you are likely to want to match a union argument against a union formal parameter, in order to narrow the union.)
